transcode/transcoder_caching.go

```diff
@@ -0,0 +1,65 @@
+package transcode
+
+import (
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"go.senan.xyz/gonic/iout"
+)
+
+const perm = 0644
+
+type CachingTranscoder struct {
+	cachePath  string
+	transcoder Transcoder
+}
+
+var _ Transcoder = (*CachingTranscoder)(nil)
+
+func NewCachingTranscoder(t Transcoder, cachePath string) *CachingTranscoder {
+	return &CachingTranscoder{transcoder: t, cachePath: cachePath}
+}
+
+func (t *CachingTranscoder) Transcode(ctx context.Context, profile Profile, in string) (io.ReadCloser, error) {
+	if err := os.MkdirAll(t.cachePath, perm^0111); err != nil {
+		return nil, fmt.Errorf("make cache path: %w", err)
+	}
+
+	name, args, err := parseProfile(profile, in)
+	if err != nil {
+		return nil, fmt.Errorf("split command: %w", err)
+	}
+
+	key := cacheKey(name, args)
+	path := filepath.Join(t.cachePath, key)
+
+	cf, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		return nil, fmt.Errorf("open cache file: %w", err)
+	}
+	if i, err := cf.Stat(); err == nil && i.Size() > 0 {
+		return cf, nil
+	}
+
+	out, err := t.transcoder.Transcode(ctx, profile, in)
+	if err != nil {
+		return nil, fmt.Errorf("internal transcode: %w", err)
+	}
+
+	return iout.NewTeeCloser(out, cf), nil
+}
+
+func cacheKey(cmd string, args []string) string {
+	// the cache is invalid whenever transcode command (which includes the
+	// absolute filepath, bit rate args, replay gain args, etc.) changes
+	sum := md5.New()
```

Comment:
We could use `xxhash` here, since it seems to be faster than MD5 (I used it as a hash key for cached tracks in my original PR, IIRC, and it's still here in dependency list).

[Here's](https://github.com/flashmob/go-guerrilla/issues/169) an issue in `go-guerrilla` project with some "xxhash vs MD5" benchmarks.

Reply:
yeah i think xxhash is very cool. but in this case we're dealing with really small data, and I like to keep the 3rd party libraries miminal.  on my laptop it seems to be 20µs to hash a profile and args


![20220520_204159](https://user-images.githubusercontent.com/6832539/169600450-c69f862b-3a9d-48c3-b7f7-4f9fafb2a56f.png)

Reply:
Fair enough. Well, if the `xxhash` module is not used anywhere else (probably not), then MD5 is a good choice. A bit vintage, but it works! :)